common/src/backends/metrics.rs

```diff
@@ -33,3 +41,279 @@ where
+        tower_http::trace::TraceLayer::new_for_http()
+            .make_span_with(MakeSpan::new(self.fn_span))
+            .on_response(OnResponseStatusCode)
+    }
+}
+
+impl TraceLayer<MakeSpanSimple> {
+    /// Switch to the span maker which does not add propagation details from the request headers
+    pub fn without_propagation(self) -> Self {
+        self
+    }
+}
+
+impl TraceLayer<MakeSpanPropagation> {
+    /// Switch to the span maker which adds propagation details from the request headers
+    pub fn with_propagation(self) -> Self {
+        self
+    }
+}
+
+/// Helper trait to make a new span maker
+pub trait MakeSpanBuilder {
+    fn new(fn_span: FnSpan) -> Self;
+}
+
+/// Simple span maker which records the span given by the user
+#[derive(Clone)]
+pub struct MakeSpanSimple {
+    fn_span: FnSpan,
+}
+
+impl MakeSpanBuilder for MakeSpanSimple {
+    fn new(fn_span: FnSpan) -> Self {
+        Self { fn_span }
+    }
+}
+
+impl tower_http::trace::MakeSpan<Body> for MakeSpanSimple {
+    fn make_span(&mut self, request: &Request<Body>) -> Span {
+        (self.fn_span)(request)
+    }
+}
+
+/// Span maker which records the span given by the user and extracts a propagation context
+/// from the request headers.
+#[derive(Clone)]
+pub struct MakeSpanPropagation {
+    fn_span: FnSpan,
+}
+
+impl MakeSpanBuilder for MakeSpanPropagation {
+    fn new(fn_span: FnSpan) -> Self {
+        Self { fn_span }
+    }
+}
+
+impl tower_http::trace::MakeSpan<Body> for MakeSpanPropagation {
+    fn make_span(&mut self, request: &Request<Body>) -> Span {
+        let span = (self.fn_span)(request);
+
+        let parent_context = global::get_text_map_propagator(|propagator| {
+            propagator.extract(&HeaderExtractor(request.headers()))
+        });
+        span.set_parent(parent_context);
+
+        span
+    }
+}
+
+/// Extract and records the status code from the response. And logs out timing info
+#[derive(Clone)]
+pub struct OnResponseStatusCode;
+
+impl tower_http::trace::OnResponse<BoxBody> for OnResponseStatusCode {
+    fn on_response(self, response: &Response<BoxBody>, latency: Duration, span: &Span) {
+        span.record("http.status_code", response.status().as_u16());
+        debug!(
+            latency = format_args!("{} ns", latency.as_nanos()),
+            "finished processing request"
+        );
+    }
+}
+
+/// Simple macro to record the following defaults for each request:
+/// - The URI
+/// - The method
+/// - The status code
+/// - The request path
+#[macro_export]
+macro_rules! request_span {
+    ($request:expr, $($field:tt)*) => {
+        {
+        let path = if let Some(path) = $request.extensions().get::<axum::extract::MatchedPath>() {
+            path.as_str()
+        } else {
+            ""
+        };
+
+        tracing::debug_span!(
+            "request",
+            http.uri = %$request.uri(),
+            http.method = %$request.method(),
+            http.status_code = tracing::field::Empty,
```

Comment:
Doesn't the comment for this macro indicate we will record the http status?

Reply:
Yip, however we'll only know the status once the request is complete. This happens on the [OnResponseStatusCode](https://github.com/shuttle-hq/shuttle/pull/632/files#diff-f4cf651c8b021e33d0f88c58c9b6648ff6db38f596dbb9130eaa5163a8fd4880R159)